File: api/src/utils/ebook.js
import { config } from '../config.js'
import path from 'path'

export function fileUrl(relativePath) {
  if (!relativePath) return ''
  if (relativePath.startsWith('http')) return relativePath
  const normalized = relativePath.startsWith('/') ? relativePath : `/${relativePath}`
  return `${config.appUrl}${normalized}`
}

export function formatCategory(row) {
  return {
    id: row.id,
    name: row.name,
    slug: row.slug,
    description: row.description || undefined,
    image_url: row.image_url ? fileUrl(row.image_url) : undefined,
  }
}

export function formatEbook(row, categories = []) {
  const coverPath = (row.cover_image_path || '').trim() || null
  const hasR2Cover = Boolean(row.cover_object_key)
  const hasLocalCover = Boolean(coverPath)

  // Chemin relatif : le front/mobile préfixe avec la bonne base API
  // (évite APP_URL=localhost en prod qui casse les <img>).
  let coverImageUrl = null
  if (hasR2Cover || hasLocalCover) {
    coverImageUrl = `/api/ebooks/${row.id}/cover`
  }

  return {
    id: row.id,
    title: row.title,
    slug: row.slug,
    author: row.author,
    description: row.description,
    isbn: row.isbn || undefined,
    cover_image_url: coverImageUrl,
    // Ne plus exposer le chemin PDF brut (sécurité)
    pdf_file_size: Number(row.pdf_file_size || 0),
    total_pages: row.total_pages,
    preview_pages: row.preview_pages,
    published_at: row.published_at
      ? new Date(row.published_at).toISOString().slice(0, 10)
      : null,
    is_featured: row.is_featured,
    is_active: row.is_active,
    storage_provider: row.storage_provider || 'local',
    has_pdf: Boolean(row.pdf_object_key || row.pdf_file_path),
    has_epub: Boolean(row.epub_object_key),
    has_preview: Boolean(row.preview_pdf_object_key || row.preview_epub_object_key),
    has_cover: hasR2Cover || hasLocalCover,
    categories,
    ...(row.total_views !== undefined ? { total_views: row.total_views } : {}),
  }
}

export async function loadCategoriesForEbooks(query, ebookIds) {
  if (!ebookIds.length) return new Map()
  const { rows } = await query(
    `SELECT ec.ebook_id, c.id, c.name, c.slug, c.description, c.image_url
     FROM ebook_category ec
     JOIN categories c ON c.id = ec.category_id
     WHERE ec.ebook_id = ANY($1)
     ORDER BY c.name`,
    [ebookIds]
  )
  const map = new Map()
  for (const row of rows) {
    if (!map.has(row.ebook_id)) map.set(row.ebook_id, [])
    map.get(row.ebook_id).push(formatCategory(row))
  }
  return map
}

export function validationError(res, errors, message = 'Erreur de validation') {
  return res.status(422).json({ message, errors })
}

export function parseCategoryIds(raw) {
  if (!raw) return []
  if (Array.isArray(raw)) return raw.map(Number).filter(Boolean)
  try {
    const parsed = JSON.parse(raw)
    return Array.isArray(parsed) ? parsed.map(Number).filter(Boolean) : []
  } catch {
    return []
  }
}

export const ALLOWED_IMAGE_TYPES = ['image/jpeg', 'image/jpg', 'image/png']
export const ALLOWED_PDF_TYPE = 'application/pdf'

export function extFromMimetype(mimetype) {
  const map = {
    'image/jpeg': '.jpg',
    'image/jpg': '.jpg',
    'image/png': '.png',
    'application/pdf': '.pdf',
  }
  return map[mimetype] || path.extname(mimetype)
}

File: mobile/lib/shared/models/ebook.dart
import 'category.dart';

class Ebook {
  const Ebook({
    required this.id,
    required this.title,
    required this.slug,
    required this.author,
    required this.description,
    required this.coverImageUrl,
    required this.totalPages,
    required this.previewPages,
    this.isbn,
    this.pdfFileSize = 0,
    this.publishedAt,
    this.isFeatured = false,
    this.isActive = true,
    this.categories = const [],
    this.totalViews,
  });

  final int id;
  final String title;
  final String slug;
  final String author;
  final String description;
  final String? isbn;
  final String coverImageUrl;
  final int pdfFileSize;
  final int totalPages;
  final int previewPages;
  final String? publishedAt;
  final bool isFeatured;
  final bool isActive;
  final List<Category> categories;
  final int? totalViews;

  factory Ebook.fromJson(Map<String, dynamic> json) {
    final cats = (json['categories'] as List<dynamic>? ?? [])
        .whereType<Map<String, dynamic>>()
        .map(Category.fromJson)
        .toList();

    return Ebook(
      id: json['id'] as int,
      title: json['title'] as String? ?? '',
      slug: json['slug'] as String? ?? '',
      author: json['author'] as String? ?? '',
      description: json['description'] as String? ?? '',
      isbn: json['isbn'] as String?,
      coverImageUrl: json['cover_image_url'] as String? ?? '',
      pdfFileSize: (json['pdf_file_size'] as num?)?.toInt() ?? 0,
      totalPages: (json['total_pages'] as num?)?.toInt() ?? 0,
      previewPages: (json['preview_pages'] as num?)?.toInt() ?? 10,
      publishedAt: json['published_at'] as String?,
      isFeatured: json['is_featured'] == true,
      isActive: json['is_active'] != false,
      categories: cats,
      totalViews: (json['total_views'] as num?)?.toInt(),
    );
  }
}

class EbooksPage {
  const EbooksPage({
    required this.data,
    required this.currentPage,
    required this.lastPage,
    required this.perPage,
    required this.total,
  });

  final List<Ebook> data;
  final int currentPage;
  final int lastPage;
  final int perPage;
  final int total;

  factory EbooksPage.fromJson(Map<String, dynamic> json) {
    final items = (json['data'] as List<dynamic>? ?? [])
        .whereType<Map<String, dynamic>>()
        .map(Ebook.fromJson)
        .toList();

    return EbooksPage(
      data: items,
      currentPage: (json['current_page'] as num?)?.toInt() ?? 1,
      lastPage: (json['last_page'] as num?)?.toInt() ?? 1,
      perPage: (json['per_page'] as num?)?.toInt() ?? 20,
      total: (json['total'] as num?)?.toInt() ?? items.length,
    );
  }
}
